packages/discovery/src/discovery/handlers/user/DynamicArrayHandler.ts

```diff
@@ -0,0 +1,133 @@
+import { assert } from '@l2beat/backend-tools'
+import { utils } from 'ethers'
+import * as z from 'zod'
+
+import { Bytes } from '../../../utils/Bytes'
+import { EthereumAddress } from '../../../utils/EthereumAddress'
+import { getErrorMessage } from '../../../utils/getErrorMessage'
+import { DiscoveryLogger } from '../../DiscoveryLogger'
+import { DiscoveryProvider } from '../../provider/DiscoveryProvider'
+import { Handler, HandlerResult } from '../Handler'
+import { getReferencedName, Reference, resolveReference } from '../reference'
+import { BytesFromString, NumberFromString } from '../types'
+import { bytes32ToContractValue } from '../utils/bytes32ToContractValue'
+import { valueToBigInt } from '../utils/valueToBigInt'
+
+const SingleSlot = z.union([
+  z.number().int().nonnegative(),
+  BytesFromString,
+  NumberFromString,
+  Reference,
+])
+
+export type DynamicArrayHandlerDefinition = z.infer<
+  typeof DynamicArrayHandlerDefinition
+>
+export const DynamicArrayHandlerDefinition = z.strictObject({
+  type: z.literal('dynamicArray'),
+  slot: SingleSlot,
+  returnType: z.optional(z.enum(['address'])),
+  ignoreRelative: z.optional(z.boolean()),
+})
+
+export class DynamicArrayHandler implements Handler {
```

Comment:
To keep things clear and consistent (since we already have `StorageHandler`), what do you think about renaming to `StorageArrayHandler`?